### PR TITLE
[SPARK-20656][CORE]Support Incremental parsing of event logs in SHS

### PR DESCRIPTION
**Pull request by shahidki31**
### What changes were proposed in this pull request?
Currently loading application page from history server is time consuming, if the eventlog size is high. (Eg: ~47 minutes for a 18GB eventlog file). Currently, when there is any changes in event log, history server parses the entire eventLog even for smaller changes in eventLog. In this PR, we are supporting incremental parsing of event log by storing the in memory store to a hash map and will not close the store until it is valid.

### Why are the changes needed?
To speed up loading history server page

### Does this PR introduce any user-facing change?
No

### How was this patch tested?
Added UT, Manual tests and existing tests.

Created an eventLog of size ~2GB. Added a few more events. Loading time without the PR is ~1 minute and after the PR, it is around 2 secs.

### CODE REVIEW

**Comment by shahidki31** [comment]
cc @vanzin @squito Kindly review

**Comment by dongjoon-hyun** [comment]
cc @HeartSaVioR

**Comment by HeartSaVioR** [comment]
Thanks for cc.ing me, @dongjoon-hyun . I'll take a look.

Btw, I think we have another JIRA issue for supporting incremental parsing [SPARK-28870](https://issues.apache.org/jira/browse/SPARK-28870) which has broader goal - run with any implementation of KVStore.

At first glance, this patch could cover SPARK-29261 and with SPARK-29111 it may resolve SPARK-28870 altogether - though we struggled on the details previously so I need some time to go through deeply.

@shahidki31 
I guess you've been following through the previous discussions/efforts @Ngone51 and me, and @vanzin, @squito have been made. (#25577 and #25943, and relevant google docs in relevant JIRA issues/PRs) If not, it would worth to go through, as we've discussed in details.

**Comment by shahidki31** [comment]
Thanks @HeartSaVioR , I will go through it.
 It seems, I haven't used snapshot/checkpoint for storing the underlying kvstore. Instead, in this PR, I am not closing the store, whenever there are changes in event logs or invalidate cache. Currently it supports InMemory store, but we can easily extend to disk store as well.

**Comment by HeartSaVioR** [comment]
> Instead, in this PR, I am not closing the store, whenever there are changes in event logs or invalidate cache.

This patch is simpler because this doesn't take "restarting SHS" into account. Restarting SHS will lose the information. And for now we may not want to tracking line offset in SHS's KV store (`listing`) since the line offset is only effective during single run of SHS.

When you consider restoring KV store & state of listeners during restarting of SHS, you will have to store the snapshot of KV store into somewhere (that's why SPARK-29111 came in) and then you have to concern about compatibility of snapshot (entities in KV store including live entities on listeners) across Spark versions. That's why I had to change the design and introduce SPARK-29779 instead of snapshotting.

We've already gone through bunch of discussions because it is not that simple in reality as it seems; so please go through these PRs as well as design docs. 

I guess the patch can be reviewed right now if the community prefers to have a solution which works within single SHS run first (though this may conflict with compaction #26416 which needs some arrangement), but if the community wants to have a solution which covers more cases, SPARK-28870 seems to be the way to go. (It doesn't mean this patch will not be valid - this patch could cover SPARK-29261 with some modification.)

**Comment by oopDaniel** [comment]
I was also working on a PR for incremental parsing in SHS as well, and there are 2 ways to skip the parsed content: filtering by lines (the approach in this PR) and skipping by bytes (see `mergeApplicationListing`). From what I tested locally, filtering by lines roughly takes about 30 seconds for file sizes ranging from 10MB to 400MB, while skipping by bytes only takes 2 ms for a 400MB file.

**Comment by HeartSaVioR** [comment]
> there are 2 ways to skip the parsed content: filtering by lines (the approach in this PR) and skipping by bytes (see mergeApplicationListing).

Yeah if possible we should deal with latter approach. I guess that brings more changes as ReplayListener just relies on Scala API which provides lines (no offset information) so we should get our hands be dirty, but it definitely worths to do.

**Comment by shahidki31** [comment]
> From what I tested locally, filtering by lines roughly takes about 30 seconds for file sizes ranging from 10MB to 400MB, while skipping by bytes only takes 2 ms for a 400MB file.

Hi @oopDaniel , Actually I tried with both the approach, and it seems skipping bytes seems more complicated as we need to handle more edge cases. Also, I tested this PR with 2GB event log file and I think the time to load UI took around 2 seconds (?) including filtering and replaying. Also it is not difficult to add the skipping bytes, as all we need to do is add the bytes read parameter instead of lines read parameter and handle the edge cases. @HeartSaVioR I think the approach which you guys are doing is great, as it handles restarting SHS. But, if we can review this PR related to incremental parsing, extending to snapshotting would be easier I guess.. If there is a working PR for that, I can close this.

**Comment by HeartSaVioR** [comment]
@shahidki31 
No I didn't intend to persuade you to close this. I'd just wanted to make sure we get a clear picture of full implementation before dealing with each part, but it's OK for me if you'd like to deal with current solution as I think I can deal with extending the solution with snapshotting.

I could take a look with current solution, but you still need to persuade at least one committer to push this forward.

Btw, we'd be better to clarify the performance test in details. It should include at least...

* size of event log file for initial load
* elapsed time for initial load
* count/size of events for addition (mostly about size)
* elapsed time for loading additional events

For me, your statement in PR description sounds to me as skipping (via read and drop) 2G takes around 2 secs which is still not ideal (as we know how to do it better), though I agree that's still a huge improvement.

**Comment by shahidki31** [comment]
> Btw, we'd be better to clarify the performance test in details. It should include at least...
> 
> size of event log file for initial load
> elapsed time for initial load
> count/size of events for addition (mostly about size)
> elapsed time for loading additional events

Thanks @HeartSaVioR , I'll try more performance testing based on size and initial load.

File: core/src/main/scala/org/apache/spark/status/AppStatusListener.scala [hunk 1]
```diff
@@ -100,6 +102,41 @@ private[spark] class AppStatusListener(
...
+        kvstore.write(data)
+      }
+    }
+  }
+
+  def initialize(appId: String, attemptId: Option[String]): Unit = {
+    if (!live) {
+      // If incremental parsing is enabled, read and update the listener data
+      // from store
+      this.appId = appId
```
Comment (HeartSaVioR) [inline core/src/main/scala/org/apache/spark/status/AppStatusListener.scala]: I'd rather have some case class (or just simply tuple) to store the information altogether and name it which clearly representing "incremental parsing is used". Checking whether appId is available or not to determine whether incremental parsing is used or not wouldn't be easy to understand for newcomer of the code.

File: core/src/main/scala/org/apache/spark/status/storeTypes.scala [hunk 2]
```diff
@@ -443,6 +446,24 @@ private[spark] class PoolData(
...
+    val liveRDDs: mutable.HashMap[Int, LiveRDD],
+    val pools: mutable.HashMap[String, SchedulerPool],
+    val appInfo: ApplicationInfo,
+    val coresPerTask: Int,
+    val appSummary: AppSummary,
+    val activeExecutorCount: Int) {
+
+  @KVIndex @JsonIgnore
+  def key: Array[Option[String]] = Array(Some(appId), attemptId)
+}
```
Comment (HeartSaVioR) [inline core/src/main/scala/org/apache/spark/status/storeTypes.scala]: nit: empty line after this

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLAppStatusListener.scala [hunk 3]
```diff
@@ -66,6 +68,29 @@ class SQLAppStatusListener(
         // away.
         exec.metricsValues = aggregateMetrics(exec)
         exec.write(kvstore, now)
+        if (appId != null) {
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLAppStatusListener.scala]: This should not be in `foreach` statement.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLAppStatusListener.scala [hunk 4]
```diff
@@ -66,6 +68,29 @@ class SQLAppStatusListener(
...
+          kvstore.write(new SQLAppStatusListenerData(appId, attemptId,
+            liveExecutions, stageMetrics))
+        }
+      }
+    }
+  }
+
+  def initialize(appId: String, attemptId: Option[String]): Unit = {
+    if (!live) {
+      this.appId = appId
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLAppStatusListener.scala]: same here

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLAppStatusListener.scala [hunk 5]
```diff
@@ -66,6 +68,29 @@ class SQLAppStatusListener(
...
+      this.attemptId = attemptId
+      try {
+        val listenerData = kvstore.read(classOf[SQLAppStatusListenerData],
+          Array(Some(appId), attemptId))
+        listenerData.liveExecutions.entrySet().asScala.foreach { entry =>
+          liveExecutions.put(entry.getKey, entry.getValue)
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLAppStatusListener.scala]: exec.metricsValues may need to be reset to null, see comment on `kvstore.onFlush`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLAppStatusStore.scala [hunk 6]
```diff
@@ -143,3 +144,13 @@ case class SQLPlanMetric(
     name: String,
     accumulatorId: Long,
     metricType: String)
+
+class SQLAppStatusListenerData(
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLAppStatusStore.scala]: maybe better to place it to SQLAppStatusListener as other KVStore entities are there.

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/ui/HiveThriftServer2Listener.scala [hunk 7]
```diff
@@ -69,6 +72,29 @@ private[thriftserver] class HiveThriftServer2Listener(
   kvstore.onFlush {
     if (!live) {
       flush((entity: LiveEntity) => updateStoreWithTriggerEnabled(entity))
+      if (appId != null) {
```
Comment (HeartSaVioR) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/ui/HiveThriftServer2Listener.scala]: place it before flush()

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/ui/HiveThriftServer2Listener.scala [hunk 8]
```diff
@@ -69,6 +72,29 @@ private[thriftserver] class HiveThriftServer2Listener(
...
+        kvstore.write(new HiveThriftserver2ListenerData(
+          appId, attemptId, sessionList, executionList))
+      }
+    }
+  }
+
+  def initialize(appId: String, attemptId: Option[String]): Unit = {
+    if (!live) {
+      this.appId = appId
```
Comment (HeartSaVioR) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/ui/HiveThriftServer2Listener.scala]: same here

File: core/src/test/scala/org/apache/spark/deploy/history/FsHistoryProviderSuite.scala [hunk 9]
```diff
@@ -190,6 +191,54 @@ class FsHistoryProviderSuite extends SparkFunSuite with Matchers with Logging {
     provider.mergeApplicationListingCall should be (1)
   }
 
+  test("support incremental parsing of the event logs") {
```
Comment (HeartSaVioR) [inline core/src/test/scala/org/apache/spark/deploy/history/FsHistoryProviderSuite.scala]: I know most of tests in FsHistoryProviderSuite don't deal with rolling event log, but this should really deal with it, as it's not just simple sequential reading of files.

File: core/src/test/scala/org/apache/spark/deploy/history/FsHistoryProviderSuite.scala [hunk 10]
```diff
@@ -190,6 +191,54 @@ class FsHistoryProviderSuite extends SparkFunSuite with Matchers with Logging {
...
+      SparkListenerJobEnd(0, 3L, JobSucceeded)
+    )
+
+    updateAndCheck(provider) { list =>
+      store should not be null
+      store.read(classOf[AppStatusListenerData], Array(Some("app1"), None)) should not be null
+      list.size should be (1)
+      provider.getAttempt("app1", None).logPath should endWith(EventLogFileWriter.IN_PROGRESS)
+      val appUi = provider.getAppUI("app1", None)
+      appUi should not be null
```
Comment (HeartSaVioR) [inline core/src/test/scala/org/apache/spark/deploy/history/FsHistoryProviderSuite.scala]: I think it should really check whether store reflects the events correctly. It's not enough to just check it's loaded or not. That should cover these cases - initial read / new addition of events in same file / new file.

**Review by HeartSaVioR — COMMENTED** [review]
Test coverage seems to be really low. I'd expect at least tests each listener can store the information via `flush` and reload from `initialize`, or at least these entities can be serialized/deserialized well, because I was encountered some issue when I just tried to serialize entities in events.

**Comment by shahidki31** [comment]
Thanks @HeartSaVioR for the review. I will update the PR as well as post the performance reports.

**Pull request closed without merging**